### PR TITLE
scheduler: add conflict detect for grant hot leader scheduler

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Issue Number: Ref #4399

### Check List

Tests

- No code

### Release note

```release-note
 None.
```

### CODE REVIEW

**Review by bufferflies — APPROVED** [review]

File: tools/pd-ctl/pdctl/command/scheduler.go [hunk 1]
```diff
@@ -329,7 +329,7 @@ func NewLabelSchedulerCommand() *cobra.Command {
 func NewGrantHotRegionSchedulerCommand() *cobra.Command {
 	c := &cobra.Command{
 		Use:   "grant-hot-region-scheduler <store_leader_id> <store_leader_id,store_peer_id_1,store_peer_id_2>",
-		Short: "add a scheduler to grant hot region to fixed stores",
+		Short: "add a scheduler to grant hot region to fixed stores. Note: balance-hot-region-scheduler must be paused.",
```
Comment (JmPotato) [inline tools/pd-ctl/pdctl/command/scheduler.go:332]: Could we do some actual checks to prevent these two schedulers from working at the same time?

Comment (lhy1024) [reply tools/pd-ctl/pdctl/command/scheduler.go:332]: Good idea.

File: tools/pd-ctl/pdctl/command/scheduler.go [hunk 2]
```diff
@@ -354,9 +354,10 @@ func NewSplitBucketSchedulerCommand() *cobra.Command {
 // NewGrantHotRegionSchedulerCommand returns a command to add a grant-hot-region-scheduler.
 func NewGrantHotRegionSchedulerCommand() *cobra.Command {
 	c := &cobra.Command{
-		Use:   "grant-hot-region-scheduler <store_leader_id> <store_leader_id,store_peer_id_1,store_peer_id_2>",
-		Short: "add a scheduler to grant hot region to fixed stores",
-		Run:   addSchedulerForGrantHotRegionCommandFunc,
+		Use: "grant-hot-region-scheduler <store_leader_id> <store_leader_id,store_peer_id_1,store_peer_id_2>",
+		Short: `add a scheduler to grant hot region to fixed stores. 
+				Note: If there is balance-hot-region-scheduler running, please remove it first, otherwise grant-hot-region-scheduler will not work.`,
```
Comment (bufferflies) [inline tools/pd-ctl/pdctl/command/scheduler.go:359]: is it ok if the hot scheduler is not work but exist, for example being pause status?

Comment (lhy1024) [reply tools/pd-ctl/pdctl/command/scheduler.go:359]: In this pr, even though the hot-scheduler is paused, adding grant-hot-leader-scheduler is still not allowed.

If this is allowed, let's look at an example where I suspend the hot-scheduler for ten seconds and then add grant-hot-leader-scheduler, which is fine for ten seconds, but after ten seconds the hot-scheduler resumes scheduling and they will conflict again.

Comment (bufferflies) [reply tools/pd-ctl/pdctl/command/scheduler.go:359]: Ok, understand. BTW, the hot scheduler config still exist if I enable it again.

Comment (lhy1024) [reply tools/pd-ctl/pdctl/command/scheduler.go:359]: yes,

Comment (lhy1024) [reply tools/pd-ctl/pdctl/command/scheduler.go:359]: Yes, the configuration still exists. If you think this pr is ok, I will cancel /hold.

**Comment by bufferflies** [comment]
LGTM

File: tools/pd-ctl/tests/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -488,12 +488,32 @@ func (suite *schedulerTestSuite) checkSchedulerConfig(cluster *pdTests.TestClust
 	})
 
 	// test grant hot region scheduler config
-	checkSchedulerCommand(re, cmd, pdAddr, []string{"-u", pdAddr, "scheduler", "add", "grant-hot-region-scheduler", "1", "1,2,3"}, map[string]bool{
+	// case 1: add grant-hot-region-scheduler when balance-hot-region-scheduler is running
```
Comment (rleungx) [inline tools/pd-ctl/tests/scheduler/scheduler_test.go:491]: It should not belong to the scheduler config test.

File: server/api/scheduler.go [hunk 4]
```diff
@@ -246,6 +270,48 @@ func (h *schedulerHandler) PauseOrResumeScheduler(w http.ResponseWriter, r *http
...
+	}
+	if !rc.IsServiceIndependent(constant.SchedulingServiceName) {
+		schedulers, err := h.GetSchedulerByStatus("", false)
+		if err != nil {
+			return nil, err
+		}
+		return schedulers.([]string), nil
+	}
+
+	addr, ok := h.svr.GetServicePrimaryAddr(h.svr.Context(), constant.SchedulingServiceName)
```
Comment (rleungx) [inline server/api/scheduler.go:286]: Why do we need to query schedulers from the scheduling service?

Comment (lhy1024) [reply server/api/scheduler.go:286]: Because we need to check whether hot scheduler is disabled.

File: tools/pd-ctl/pdctl/command/scheduler.go [hunk 5]
```diff
@@ -354,9 +354,10 @@ func NewSplitBucketSchedulerCommand() *cobra.Command {
 // NewGrantHotRegionSchedulerCommand returns a command to add a grant-hot-region-scheduler.
 func NewGrantHotRegionSchedulerCommand() *cobra.Command {
 	c := &cobra.Command{
-		Use:   "grant-hot-region-scheduler <store_leader_id> <store_leader_id,store_peer_id_1,store_peer_id_2>",
-		Short: "add a scheduler to grant hot region to fixed stores",
-		Run:   addSchedulerForGrantHotRegionCommandFunc,
+		Use: "grant-hot-region-scheduler <store_leader_id> <store_leader_id,store_peer_id_1,store_peer_id_2>",
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/scheduler.go:357]: BTW, is it necessary to `set store_leader_id` again?

Comment (lhy1024) [reply tools/pd-ctl/pdctl/command/scheduler.go:357]: Sure, I also think it is not necessary. I will modify it in another pr.

File: server/api/scheduler.go (hunk 4 above)
Comment (rleungx) [reply server/api/scheduler.go:286]: PD has the same configuration as the scheduling service?

Comment (lhy1024) [reply server/api/scheduler.go]: done

**Review by rleungx — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Pull request merged**